AssertionError: Detected an inconsistency in the cache

On a particular production deployment of RS 2.1.1, using Python 2.7, gevent, and MySQL ??? (with the umysqldb driver) one server started producing a bunch of cache assertion errors. Example:

```
storage_cache.py line 432 called self._check_tid_after_load()

AssertionError: Detected an inconsistency between the RelStorage cache
and the database while loading an object using the delta_after0 dict.
Please verify the database is configured for ACID compliance and that
all clients are using the same commit lock.

oid_int=97,214,675,
# tid_int from delta_after0    -> 2019-05-10 13:08:56.434258
expect_tid_int=274,612, 602, 263,913,028, 
# actual_tid_int from db       -> 2019-05-10 13:16:50.574925
actual_tid_int=274,612, 636, 204,224,000L,
# from polling                 -> 2019-05-10 14:39:47.624924
current_tid   =274,612, 992, 475,340,270L,

cp0           =274,612, 314, 553,991,492,
cp1           =274,610, 362, 479,892,548,

len(delta_after0)=13641,
len(delta_after1)=19782,
pid=9322,
thread_ident=<_DummyThread(DummyThread-27838, started daemon >)
```

Notes from my investigation follow.

(I'll abbreviate the tids to their middle group of digits.)

`expect_tid_int` means that for object 97,214,675, this connection has information that the last time it changed was in transaction 602 (13:08:56 UTC). It has this information either because it itself was the one that changed the object in that transaction, or, at some point in the past it polled the database for changes and was told that somebody else changed the object in that transaction.

`current_tid` is the last transaction recorded in the database when this connection polled for invalidations. We poll for invalidations every time we open a new transaction (and indeed, every time we finish a transaction). This poll happened at 14:39.

When we went to request the current data from the database, we got back state information for that object as-of transaction 636 (`actual_tid_int` is 13:16:50 --- almost eight minutes later). That doesn't match what we were expecting, so *somebody* changed the object without us finding out. (This data did *not* come from a cache, it came from the database.) Moreover, this doesn't make any sense: since we polled at 14:39, we *should* have found the information that told us that the object was changed at 13:16, and dropped the idea that it was changed at 13:08.

How could we get in this state? Good question. 

- There could be some sort of bug in the way that relstorage is handling polling for invalidations and updating its idea of what the current state of the objects should be. The logic is complicated (I can't keep it in my head, I have to read it carefully every time I look at it) so I can't rule that out. I think it can ultimately be simplified.

- There could have been a long running transaction that committed old data *after* we polled for invalidations. That's not supposed to happen if the commit locks are working as they're supposed to ("Please verify...that all clients are using the same commit lock.") There's no choice of configuring the lock under MySQL, it always locks using "SELECT GET_LOCK(CONCAT(DATABASE(), '.commit'), timeout)". Unless we've got some weird multi-master MySQL setup or something such that locks aren't actually shared, I don't know how that could be wrong. 

But that circles us back to umysqldb and some of the mysterious errors we've seen there. Here's a snippet from the lock code:

```python
stmt = "SELECT GET_LOCK(CONCAT(DATABASE(), '.commit'), %s)"
cursor.execute(stmt, (timeout,))
try:
    locked = cursor.fetchone()[0]
except TypeError as e: # pragma: no cover
    # This has been observed under certain database drivers and concurrency loads,
    # specifically gevent with umysqldb and high concurrency. It's not clear what the cause
    # is, so lets at least raise a specific message.
    # It's often TypeError("'NoneType' object has no attribute '__getitem__'",),
    # meaning that cursor.fetchone() returned nothing. This may be related to
    # cursor.lastrowid being None in OID allocator sometimes.
    raise CommitLockQueryFailedError("The commit lock query failed: %s" % repr(e))
```
So I wonder if we *think* we're locking, but we're actually not due to a driver issue (an error passing silently)? 

Also, what MySQL version are we on? I also have documented weirdness with <= 5.7.5 and locks.

- MVCC issues in MySQL (not getting a consistent table snapshot)?

If it's either of the first two problems, restarting the instance will probably take care of it.

I think RelStorage could probably do something "better" than raising `AssertionError` here. It could maybe raise an exception that causes the cache to clear itself and then the transaction to get retried. If there's an underlying bug in the polling, that would effectively mask it. I put "better" in quotes because I'm not sure that it really is better or not. Thoughts?

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -84,7 +84,7 @@
   happen we now reset the cache and raise a type of ``TransientError``
   allowing the application to retry. A few instances where previously
   incorrect data could be cached may now raise such a
-  ``TransientError``.
+  ``TransientError``. See :pr:`245`.
 
 2.1.1 (2019-01-07)
 ==================

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -75,6 +75,17 @@
   1-2% before). It is currently slower to read and write, however.
   This is a work in progress. See :pr:`243`.
 
+- Add more aggressive validation and, when possible, corrections for
+  certain types of cache consistency errors. Previously an
+  ``AssertionError`` would be raised with the message "Detected an
+  inconsistency between RelStorage and the database...". We now
+  proactively try harder to avoid that situation based on some
+  educated guesses about when it could happen, and should it still
+  happen we now reset the cache and raise a type of ``TransientError``
+  allowing the application to retry. A few instances where previously
+  incorrect data could be cached may now raise such a
+  ``TransientError``.
+
 2.1.1 (2019-01-07)
 ==================
 

--- a/src/relstorage/adapters/interfaces.py
+++ b/src/relstorage/adapters/interfaces.py
@@ -542,10 +542,16 @@ class IPoller(Interface):
         """
 
     def list_changes(cursor, after_tid, last_tid):
-        """Return the (oid, tid) values changed in a range of transactions.
+        """
+        Return the ``(oid, tid)`` values changed in a range of
+        transactions.
 
-        The returned iterable must include all changes in the range
-        after_tid < tid <= last_tid.
+        The returned sequence (which has a defined ``len``) must
+        include the latest changes in the range *after_tid* < ``tid``
+        <= *last_tid*.
+
+        The ``oid`` values returned will be distinct: each ``oid``
+        will have been changed in exactly one ``tid``.
         """
 
 

--- a/src/relstorage/adapters/poller.py
+++ b/src/relstorage/adapters/poller.py
@@ -29,6 +29,9 @@ log = logging.getLogger(__name__)
 class Poller(object):
     """Database change notification poller"""
 
+    # The zoid is the primary key on both ``current_object`` (history
+    # preserving) and ``object_state`` (history free), so these
+    # queries are guaranteed to only produce an OID once.
     _list_changes_range_queries = (
         """
         SELECT zoid, tid
@@ -174,10 +177,8 @@ class Poller(object):
         return changes, new_polled_tid
 
     def list_changes(self, cursor, after_tid, last_tid):
-        """Return the (oid, tid) values changed in a range of transactions.
-
-        The returned iterable must include the latest changes in the range
-        after_tid < tid <= last_tid.
+        """
+        See ``IPoller``.
         """
         params = {'min_tid': after_tid, 'max_tid': last_tid}
         cursor.execute(self._list_changes_range_query, params)

--- a/src/relstorage/cache/interfaces.py
+++ b/src/relstorage/cache/interfaces.py
@@ -19,6 +19,8 @@ from zope.interface import Attribute
 from zope.interface import Interface
 
 import BTrees
+from transaction.interfaces import TransientError
+from ZODB.POSException import StorageError
 
 from relstorage._compat import PYPY
 
@@ -319,9 +321,19 @@ class IGenerationalLRUCache(ILRUCache):
     generations = Attribute("Ordered list of generations, with 0 being NoSuchGeneration.")
 
 
-class CacheCorruptedError(AssertionError):
+class CacheCorruptedError(StorageError):
     """
-    Raised when we detect cache corruption.
+    Raised when we detect internal cache corruption,
+    outside of any particular transaction.
+    """
+
+class CacheConsistencyError(StorageError, TransientError):
+    """
+    Raised when we detect internal cache corruption, having to do with
+    transactional consistency.
+
+    This is probably a transient error. By the time it is raised, the
+    faulty cache will have been cleared, and we can try again.
     """
 
 class NoSuchGeneration(object):

--- a/src/relstorage/cache/local_client.py
+++ b/src/relstorage/cache/local_client.py
@@ -242,6 +242,9 @@ class LocalClient(object):
 
     def store_checkpoints(self, cp0, cp1):
         # No lock, the assignment should be atomic
+        # Both checkpoints should be None, or the same integer,
+        # or cp0 should be ahead of cp1. Anything else is a bug.
+        assert cp0 == cp1 or cp0 > cp1
         cp = self.checkpoints = cp0, cp1
         return cp
 

--- a/src/relstorage/cache/storage_cache.py
+++ b/src/relstorage/cache/storage_cache.py
@@ -168,6 +168,10 @@ class StorageCache(object):
         # Also, if there have been enough changes that someone has shifted the
         # checkpoints, cache.checkpoints won't match the global checkpoints
         # and they will wind up discarding the delta maps on the first poll.
+        #
+        # Alternately, we could watch our children created here, and see
+        # which one is still alive and has the highest `current_tid` indicating the
+        # most recent poll, and copy that information.
         cache.checkpoints = self.checkpoints
         cache.delta_after0 = self._delta_map_type(self.delta_after0)
         cache.delta_after1 = self._delta_map_type(self.delta_after1)
@@ -415,7 +419,7 @@ class StorageCache(object):
             cache_data = cache[key]
             if cache_data:
                 # Cache hit.
-                assert cache_data[1] == tid_int
+                assert cache_data[1] == tid_int, cache_data[1]
                 return cache_data
 
             # Cache miss.
@@ -497,11 +501,11 @@ class StorageCache(object):
         # Under what circumstances would we get here (after commiting
         # a transaction) without ever having polled to establish
         # checkpoints? Turns out that database-level APIs like
-        # db.undo() use the master storage, not an MVCC instance. And
-        # the master storage and storage instance never gets to poll
-        # in those cases.
+        # db.undo() use new storage instances in an unusual way, and
+        # will not necessarily have polled by the time they commit.
+        #
         # Of course, if we restored from persistent cache files the master
-        # could have checkpoints.
+        # could have checkpoints we copied down.
         #
         # TODO: Create a special subclass for MVCC instances and separate
         # the state handling.

--- a/src/relstorage/cache/storage_cache.py
+++ b/src/relstorage/cache/storage_cache.py
@@ -31,6 +31,7 @@ from relstorage.cache import persistence
 from relstorage.cache.interfaces import IPersistentCache
 from relstorage.cache.interfaces import OID_TID_MAP_TYPE
 from relstorage.cache.interfaces import OID_OBJECT_MAP_TYPE
+from relstorage.cache.interfaces import CacheConsistencyError
 from relstorage.cache.local_client import LocalClient
 from relstorage.cache.memcache_client import MemcacheStateCache
 from relstorage.cache.trace import ZEOTracer
@@ -240,30 +241,48 @@ class StorageCache(object):
             self._tracer.close()
             del self._tracer
 
+    def _reset(self, message=None):
+        """
+        Reset the transaction state of only this instance.
+
+        If this is being done in a transactional way, it must be followed
+        by raising an exception. If the *message* parameter is provided,
+        then a ``CacheConsistencyError`` will be raised when this
+        method returns.
+        """
+        # As if we've never polled
+        self.checkpoints = None
+        self.current_tid = 0
+        self.delta_after0 = self._delta_map_type()
+        self.delta_after1 = self._delta_map_type()
+        if message:
+            raise CacheConsistencyError(message)
+
+
     def clear(self, load_persistent=True):
         """
-        Remove all data from the cache.  Called by speed tests.
+        Remove all data from the cache, both locally (and shared among
+        other instances), and globally.
 
-        Starting from the introduction of persistent cache files,
-        this also results in the local client being repopulated with
-        the current set of persistent data. The *load_persistent* keyword can
-        be used to control this.
+        Called by speed tests.
 
-        .. versionchanged:: 2.0b6
-           Added the ``load_persistent`` keyword. This argument is provisional.
+        Starting from the introduction of persistent cache files, this
+        also results in the local client being repopulated with the
+        current set of persistent data. The *load_persistent* keyword
+        can be used to control this.
+
+        .. versionchanged:: 2.0b6 Added the ``load_persistent``
+        keyword. This argument is provisional.
         """
-        self.cache.flush_all()
-
-        self.checkpoints = None
+        self._reset()
         # After this our current_tid is probably out of sync with the
         # storage's current_tid. Whether or not we load data from
         # persistent caches, it's probably in the past of what the
         # storage thinks.
         # XXX: Ideally, we should be able to populate that information
         # back up so that we get the right polls.
-        self.current_tid = 0
-        self.delta_after0 = self._delta_map_type()
-        self.delta_after1 = self._delta_map_type()
+
+        self.cache.flush_all()
 
         if load_persistent:
             self.restore()
@@ -291,6 +310,10 @@ class StorageCache(object):
 
         if expect_tid_int is not None and actual_tid_int != expect_tid_int:
             # Uh-oh, the cache is inconsistent with the database.
+            # We didn't get a TID from the future, but it's not what we
+            # had in our delta_after0 map, which means...we missed a change
+            # somewhere.
+            #
             # Possible causes:
             #
             # - The database MUST provide a snapshot view for each
@@ -300,6 +323,10 @@ class StorageCache(object):
             #   can not provide a snapshot view. (InnoDB is
             #   sufficient.)
             #
+            # - (Similar to the last one.) Using too low of a
+            #   isolation level for the database connection and
+            #   viewing unrelated data.
+            #
             # - Something could be writing to the database out
             #   of order, such as a version of RelStorage that
             #   acquires a different commit lock.
@@ -308,6 +335,10 @@ class StorageCache(object):
             #   in after_poll() that caused it to ignore the
             #   transaction order, leading it to sometimes put the
             #   wrong tid in delta_after*.
+            #
+            # - Restarting a load connection at a future point we hadn't
+            #   actually polled to, such that our current_tid is out of sync
+            #   with the connection's *actual* viewable tid?
             cp0, cp1 = self.checkpoints
 
             msg = ("Detected an inconsistency "
@@ -333,8 +364,9 @@ class StorageCache(object):
                        'pid': os.getpid(),
                        'thread_ident': threading.current_thread(),
                    })
-            # TODO: Make this rasie a CacheCorruptedError.
-            raise AssertionError(msg)
+            # We reset ourself as if we hadn't polled, and hope the transient
+            # error gets retried in a working, consistent view.
+            self._reset(msg)
 
     def load(self, cursor, oid_int):
         """
@@ -363,12 +395,17 @@ class StorageCache(object):
         # checkpoints[0] is the preferred location.
         #
         # If delta_after0 contains oid_int, we should not look at any
-        # other cache keys, since the tid_int specified in delta_after0
-        # replaces all older transaction IDs. Similarly, if
-        # delta_after1 contains oid_int, we should not look at
-        # checkpoints[1]. Also, when both checkpoints are set to the
-        # same transaction ID, we don't need to ask for the same key
-        # twice.
+        # other cache keys, since the tid_int specified in
+        # delta_after0 replaces all older transaction IDs. We *know*
+        # that oid_int should be at (exactly) tid_int because we
+        # either made that change ourself (after_tpc_finish) *or* we
+        # have polled within our current database transaction (or a
+        # previous one) and been told that the oid changed in tid.
+        #
+        # Similarly, if delta_after1 contains oid_int, we should not
+        # look at checkpoints[1]. Also, when both checkpoints are set
+        # to the same transaction ID, we don't need to ask for the
+        # same key twice.
         cache = self.cache
         tid_int = self.delta_after0.get(oid_int)
         if tid_int:
@@ -427,11 +464,14 @@ class StorageCache(object):
         self.read_temp = q.read_temp
 
     def _send_queue(self, tid):
-        """Now that this tid is known, send all queued objects to the cache"""
+        """
+        Now that this tid is known, send all queued objects to the
+        cache. The cache will have ``(oid, tid)`` entry for each object
+        we have been holding on to (well, in a big transaction, some of them
+        might actually not get stored in the cache. But we try!)
+        """
         tid_int = u64(tid)
 
-        # Trace these. This is the equivalent of ZEO's
-        # ClientStorage._update_cache.
         self.cache.set_all_for_tid(tid_int, self.queue)
         # We only do this because cache_trace_analysis uses us
         # in ways that aren't quite accurate. We'd prefer to call clear_temp()
@@ -451,8 +491,20 @@ class StorageCache(object):
         if self.checkpoints:
             for oid_int in self.queue.stored_oids:
                 # Future cache lookups for oid_int should now use
-                # the tid just committed.
+                # the tid just committed. We're about to flush that
+                # data to the cache.
                 self.delta_after0[oid_int] = tid_int
+        # Under what circumstances would we get here (after commiting
+        # a transaction) without ever having polled to establish
+        # checkpoints? Turns out that database-level APIs like
+        # db.undo() use the master storage, not an MVCC instance. And
+        # the master storage and storage instance never gets to poll
+        # in those cases.
+        # Of course, if we restored from persistent cache files the master
+        # could have checkpoints.
+        #
+        # TODO: Create a special subclass for MVCC instances and separate
+        # the state handling.
 
         self._send_queue(tid)
 
@@ -580,42 +632,55 @@ class StorageCache(object):
         # We have to replace the checkpoints.
         cp0, cp1 = new_checkpoints
 
-        # Use the checkpoints specified by the cache.
-        # Rebuild delta_after0 and delta_after1.
+        # Use the checkpoints specified by the cache (or equal to new_tid_int,
+        # if the cache was in the future.)
+
+        # Rebuild delta_after0 and delta_after1, if we can.
+        # If we can't, because we don't actually have a range, do nothing.
+        # If the case that the checkpoints are (new_tid, new_tid),
+        # we'll do nothing and have no delta maps. This is because, hopefully,
+        # next time we poll we'll be able to use the global checkpoints and
+        # catch up then.
         new_delta_after0 = self._delta_map_type()
         new_delta_after1 = self._delta_map_type()
         if cp1 < new_tid_int:
-            # poller.list_changes(after_tid, last_tid) provides an iterator of
-            # (oid, tid) where tid > after_tid and tid <= last_tid.
+            # poller.list_changes(cp1, new_tid_int) provides an iterator of
+            # (oid, tid) where tid > cp1 and tid <= new_tid_int. It is guaranteed
+            # that each oid shows up only once.
             change_list = self.adapter.poller.list_changes(
                 cursor, cp1, new_tid_int)
-
-            # Make a dictionary that contains, for each oid, the most
-            # recent tid listed in changes. This works because sorting the
-            # (oid, tid) pairs puts the newest tid at the back, and constructing
-            # the dictionary from that sorted list preserves order, keeping the
-            # last key that it saw.
-            change_list = sorted(change_list) # Note this materializes the iterator
-            try:
-                change_dict = self._delta_map_type(change_list)
-            except TypeError:
-                # pg8000 returns a list of lists, not a list of tuples. The
-                # BTree constructor is very particular about that. Normally one
-                # would use pg8000 on PyPy, where we don't use BTrees, so this shouldn't
-                # actually come up in practice.
-                change_dict = self._delta_map_type()
-                for oid_int, tid_int in change_list:
-                    change_dict[oid_int] = tid_int
-            del change_list
 
             # Put the changes in new_delta_after*.
             updating_0 = self.cache.updating_delta_map(new_delta_after0)
             updating_1 = self.cache.updating_delta_map(new_delta_after1)
-            for oid_int, tid_int in change_dict.items():
+            for oid_int, tid_int in change_list:
+                if tid_int <= cp1 or tid_int > new_tid_int:
+                    self._reset(
+                        "Requested changes %d < tid <= %d "
+                        "but change %d for OID %d out of range." % (
+                            cp1, new_tid_int,
+                            tid_int, oid_int
+                        )
+                    )
+
                 if tid_int > cp0:
                     updating_0[oid_int] = tid_int
-                elif tid_int > cp1:
+                else:
+                    # This must be > cp1
                     updating_1[oid_int] = tid_int
+
+            # Everybody has a home (we didn't get duplicate entries
+            # or multiple entries for the same OID with different TID)
+            l0 = len(new_delta_after0)
+            l1 = len(new_delta_after1)
+            if l0 + l1 != len(change_list):
+                self._reset(
+                    "Expected delta_after0 (%d) and delta_after1 (%d) "
+                    "to have total len %d, not %d" % (
+                        l0, l1,
+                        len(change_list), l0 + l1
+                    )
+                )
 
         self.checkpoints = new_checkpoints
         self.delta_after0 = new_delta_after0

--- a/src/relstorage/cache/tests/__init__.py
+++ b/src/relstorage/cache/tests/__init__.py
@@ -50,13 +50,15 @@ class MockPoller(object):
     def __init__(self):
         self.changes = []  # [(oid, tid)]
     def list_changes(self, _cursor, after_tid, last_tid):
-        # Return a generator, because the caller shouldn't assume
-        # a length. Return exactly the item in the list because
+        # Return a list, because the caller is allowed
+        # to assume a length. Return exactly the item in the list because
         # it may be a type other than a tuple
-        for change in self.changes:
-            _, tid = change
-            if tid > after_tid and tid <= last_tid:
-                yield change
+        return [
+            item
+            for item in self.changes
+            if item[1] > after_tid and item[1] <= last_tid
+        ]
+
 
 class Cache(_BaseCache):
     # Tweak the generation sizes to match what we developed the tests with

--- a/src/relstorage/cache/tests/test_storage_cache.py
+++ b/src/relstorage/cache/tests/test_storage_cache.py
@@ -194,14 +194,16 @@ class StorageCacheTests(TestCase):
         self.assertEqual(res, (b'abc', 55))
 
     def test_load_using_delta_after0_inconsistent(self):
+        from relstorage.cache.interfaces import CacheConsistencyError
         adapter = MockAdapter()
         c = self.getClass()(adapter, MockOptionsWithFakeCache(), 'myprefix')
         c.current_tid = 60
         c.checkpoints = (50, 40)
         c.delta_after0[2] = 55
         adapter.mover.data[2] = (b'abc', 56)
-        with self.assertRaisesRegex(AssertionError, "Detected an inconsistency"):
+        with self.assertRaisesRegex(CacheConsistencyError, "Detected an inconsistency"):
             c.load(None, 2)
+        self.assertIsNone(c.checkpoints)
 
     def test_load_using_delta_after0_future_error(self):
         adapter = MockAdapter()
@@ -410,15 +412,65 @@ class StorageCacheTests(TestCase):
         self.assertEqual(dict(c.delta_after0), {2: 45})
         self.assertEqual(dict(c.delta_after1), {1: 35})
 
+    def test_after_poll_new_checkpoints_bad_changes_duplicates(self):
+        from relstorage.tests.fakecache import data
+        from relstorage.cache.interfaces import CacheConsistencyError
+        data['myprefix:checkpoints'] = b'50 40'
+
+        # Note that OID 3 changed twice. `list_changes` is required
+        # to only produce the most recent change for a given
+        # OID, so this data is in violation of the contract. Make sure
+        # that we catch that.
+        changes = [(3, 42), (1, 35), (2, 45), (3, 41)]
+
+        adapter = MockAdapter()
+        c = self.getClass()(adapter, MockOptionsWithFakeCache(), 'myprefix')
+        adapter.poller.changes = changes
+        __traceback_info__ = adapter.poller.changes
+        c.checkpoints = (40, 30)
+        c.current_tid = 40
+
+        with self.assertRaisesRegex(CacheConsistencyError, "to have total len"):
+            c.after_poll(None, 40, 50, None)
+        self.assertIsNone(c.checkpoints)
+        self.assertEqual(0, c.current_tid)
+
+    def test_after_poll_new_checkpoints_bad_changes_out_of_order(self):
+        from relstorage.tests.fakecache import data
+        from relstorage.cache.interfaces import CacheConsistencyError
+        data['myprefix:checkpoints'] = b'50 40'
+
+        adapter = MockAdapter()
+        c = self.getClass()(adapter, MockOptionsWithFakeCache(), 'myprefix')
+        c.checkpoints = (40, 30)
+        c.current_tid = 40
+
+        # Too high
+        adapter.poller.list_changes = lambda *args: [(3, 51)]
+        with self.assertRaisesRegex(CacheConsistencyError, "out of range"):
+            c.after_poll(None, 40, 50, None)
+        self.assertIsNone(c.checkpoints)
+        self.assertEqual(0, c.current_tid)
+
+        # Too low
+        c.checkpoints = (40, 30)
+        c.current_tid = 40
+        adapter.poller.list_changes = lambda *args: [(3, 40)]
+        with self.assertRaisesRegex(CacheConsistencyError, "out of range"):
+            c.after_poll(None, 40, 50, None)
+        self.assertIsNone(c.checkpoints)
+        self.assertEqual(0, c.current_tid)
+
     def test_after_poll_new_checkpoints(self):
         from relstorage.tests.fakecache import data
         data['myprefix:checkpoints'] = b'50 40'
-        # Note that OID 3 changed twice.  list_changes is not required
-        # to provide the list of changes in order, so simulate
-        # a list of changes that is out of order. It's also not required
-        # to provide a list of tuples; it could provide a list of lists.
-        # That turns out to matter to the BTree constructor.
-        changes = [(3, 42), (1, 35), (2, 45), (3, 41)]
+
+        # list_changes isn't required to provide changes in any particular
+        # order.
+        # list_changes isn't not required to provide a list of tuples;
+        # it could provide a list of lists. That turns out to matter
+        # to the BTree constructor.
+        changes = [(3, 42), (1, 35), (2, 45)]
 
         for f in (tuple, list):
             adapter = MockAdapter()

--- a/src/relstorage/storage.py
+++ b/src/relstorage/storage.py
@@ -56,6 +56,7 @@ from relstorage._compat import iterkeys
 from relstorage._compat import loads
 from relstorage.blobhelper import BlobHelper
 from relstorage.cache import StorageCache
+from relstorage.cache.interfaces import CacheConsistencyError
 from relstorage.options import Options
 
 # pylint:disable=too-many-lines
@@ -567,7 +568,12 @@ class RelStorage(UndoLogCompatible,
         with self._lock:
             self._before_load()
             cursor = self._load_cursor
-            state, tid_int = cache.load(cursor, oid_int)
+            try:
+                state, tid_int = cache.load(cursor, oid_int)
+            except CacheConsistencyError:
+                log.exception("Cache consistency error; restarting load")
+                self._drop_load_connection()
+                raise
 
         if tid_int is None:
             self._log_keyerror(oid_int, "no tid found")
@@ -1379,12 +1385,16 @@ class RelStorage(UndoLogCompatible,
             # error until the application tries to read or write something.
             # XXX: We probably need to drop our pickle cache? At least the local
             # delta_after* maps/current_tid/checkpoints?
+            log.error("ReadConflictError from polling invalidations; %s", e)
             self._stale_error = e
             return (), prev
 
         self._stale_error = None
 
         # Inform the cache of the changes.
+        # It's not good if this raises a CacheConsistencyError, that should
+        # be a fresh connection. It's not clear that we can take any steps to
+        # recover that the cache hasn't already taken.
         self._cache.after_poll(
             self._load_cursor, prev, new_polled_tid, changes)
 

--- a/src/relstorage/storage.py
+++ b/src/relstorage/storage.py
@@ -1217,6 +1217,14 @@ class RelStorage(UndoLogCompatible,
         the transaction.
         """
 
+        # This is called directly from code in DB.py on a new instance
+        # (created either by new_instance() or a special
+        # undo_instance()). That new instance is never asked to load
+        # anything, or poll invalidations, so our storage cache is ineffective
+        # (unless we had loaded persistent state files)
+        #
+        # TODO: Implement 'undo_instance' to make this clear.
+
         if self._stale_error is not None:
             raise self._stale_error
         if self._is_read_only:

